src/core/ddsi/src/q_entity.c

```diff
@@ -5414,33 +5625,51 @@ static void delete_ppt (struct proxy_participant *proxypp, ddsrt_wctime_t timest
   if (isimplicit)
     proxypp->lease_expired = 1;
 
-  /* Get snapshot of endpoints list so that we can release proxypp->e.lock
+  /* Get snapshot of endpoints and topics so that we can release proxypp->e.lock
      Pwrs/prds may be deleted during the iteration over the entities,
-     but resolving the guid will fail for these entities and the our
-     call to delete_proxy_writer/reader returns. */
+     but resolving the guid will fail for these entities and the
+     call to delete_proxy_writer/reader/topic returns. */
   {
-    eps = ddsrt_malloc (proxypp->refc * sizeof(ddsi_entityid_t));
+    child_entities = ddsrt_malloc (proxypp->refc * sizeof(ddsi_entityid_t));
     struct proxy_endpoint_common *cep = proxypp->endpoints;
     while (cep)
     {
       const struct entity_common *entc = entity_common_from_proxy_endpoint_common (cep);
-      eps[ep_count++] = entc->guid.entityid;
+      child_entities[n_child_entities++] = entc->guid.entityid;
       cep = cep->next_ep;
     }
+
+#ifdef DDS_HAS_TOPIC_DISCOVERY
+    topic_refs = ddsrt_malloc (proxy_topic_list_count (&proxypp->topics) * sizeof(struct proxy_topic *));
+    proxy_topic_list_iter_t it;
+    struct proxy_topic *proxytp = proxy_topic_list_iter_first (&proxypp->topics, &it);
+    while (proxytp != NULL)
+    {
+      topic_refs[n_topic_refs++] = proxytp;
+      proxytp = proxy_topic_list_iter_next (&it);
+    }
+#endif
   }
   ddsrt_mutex_unlock (&proxypp->e.lock);
 
   ELOGDISC (proxypp, "delete_ppt("PGUIDFMT") - deleting endpoints\n", PGUID (proxypp->e.guid));
-  ep_guid.prefix = proxypp->e.guid.prefix;
-  for (uint32_t n = 0; n < ep_count; n++)
+  ddsi_guid_t ep_guid = { .prefix = proxypp->e.guid.prefix, .entityid = { 0 } };
+  for (uint32_t n = 0; n < n_child_entities; n++)
   {
-    ep_guid.entityid = eps[n];
+    ep_guid.entityid = child_entities[n];
     if (is_writer_entityid (ep_guid.entityid))
       delete_proxy_writer (proxypp->e.gv, &ep_guid, timestamp, isimplicit);
-    else
+    else if (is_reader_entityid (ep_guid.entityid))
       delete_proxy_reader (proxypp->e.gv, &ep_guid, timestamp, isimplicit);
   }
-  ddsrt_free (eps);
+  ddsrt_free (child_entities);
+
+#ifdef DDS_HAS_TOPIC_DISCOVERY
```

Comment:
The intent here is to delete all proxy topics of this proxy participant, right? So why go through the complications of `delete_proxy_topic`, which locks the proxy participant and searches for the given definition? It's fortunate that the order of the searching is the same as the order of deleting them, otherwise it'd be O(n^2)!

I do see the locking order is a bit in the way, but it seems that it would be safe to do just delete them all earlier, where `deleting` is set. At first sight it seems that that would remove the need to build `topic_refs` as well.